bitcoin/examples/sign-tx-segwit-v0.rs

```diff
@@ -0,0 +1,125 @@
+use bitcoin::{
+    transaction, Address, Amount, Network, OutPoint, ScriptBuf, Sequence, Transaction, TxIn, TxOut,
+    Txid, WPubkeyHash, Witness,
+};
+
+const DUMMY_UTXO_AMOUNT: Amount = Amount::from_sat(20_000_000);
+const SPEND_AMOUNT: Amount = Amount::from_sat(5_000_000);
+const CHANGE_AMOUNT: Amount = Amount::from_sat(14_999_000); // 1000 sat fee.
+
+fn main() {
+    let secp = Secp256k1::new();
+
+    // Get a secret key we control and the pubkeyhash of the associated pubkey.
+    // In a real application these would come from a stored secret.
+    let (sk, wpkh) = senders_keys(&secp);
+
+    // Get an address to send to.
+    let address = receivers_address();
+
+    // Get an unspent output that is locked to the key above that we control.
+    // In a real application these would come from the chain.
+    let (dummy_out_point, dummy_utxo) = dummy_unspent_transaction_output(&wpkh);
+
+    // The input for the transaction we are constructing.
+    let input = TxIn {
+        previous_output: dummy_out_point, // The dummy output we are spending.
+        script_sig: ScriptBuf::default(), // For a p2wpkh script_sig is empty.
+        sequence: Sequence::ENABLE_RBF_NO_LOCKTIME,
+        witness: Witness::default(), // Filled in after signing.
+    };
+
+    // The spend output is locked to a key controlled by the receiver.
+    let spend = TxOut { value: SPEND_AMOUNT, script_pubkey: address.script_pubkey() };
+
+    // The change output is locked to a key controlled by us.
+    let change = TxOut {
+        value: CHANGE_AMOUNT,
+        script_pubkey: ScriptBuf::new_p2wpkh(&wpkh), // Change comes back to us.
+    };
+
+    // The transaction we want to sign and broadcast.
+    let mut unsigned_tx = Transaction {
+        version: transaction::Version::TWO,  // Post BIP-68.
+        lock_time: absolute::LockTime::ZERO, // Ignore the locktime.
+        input: vec![input],                  // Input goes into index 0.
+        output: vec![spend, change],         // Outputs, order does not matter.
+    };
+    let input_index = 0;
+
+    // Get the sighash to sign.
+    let sighash_type = EcdsaSighashType::All;
+    let mut sighasher = SighashCache::new(&mut unsigned_tx);
+    let sighash = sighasher
+        .p2wpkh_signature_hash(input_index, &dummy_utxo.script_pubkey, SPEND_AMOUNT, sighash_type)
+        .expect("failed to create sighash");
+
+    // Sign the sighash using the secp256k1 library (exported by rust-bitcoin).
+    let msg = Message::from(sighash);
+    let sig = secp.sign_ecdsa(&msg, &sk);
+
+    // Update the witness stack.
+    let signature = bitcoin::ecdsa::Signature { sig, hash_ty: EcdsaSighashType::All };
+    let pk = sk.public_key(&secp);
+    *sighasher.witness_mut(input_index).unwrap() = Witness::p2wpkh(&signature, &pk);
+
+    // Get the signed transaction.
+    let tx = sighasher.into_transaction();
+
+    // BOOM! Transaction signed and ready to broadcast.
+    println!("{:#?}", tx);
```

Comment:
Do we have guarantees that everything worked properly by this point?

Reply:
FWIW this change comes from #2095 

Do you mean do we have any API functions that can be used to guarantee this is sane before broadcasting it? Or do you mean do we have any compile time ways to check that the example is correct?

Reply:
For the first, I don't believe we do. And for the second we could do some assertions but since this is example code I think we should just get it right.